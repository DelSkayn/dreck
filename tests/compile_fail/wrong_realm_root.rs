use dreck::*;
use std::pin::pin;

pub struct Container<'gc, 'own>(Option<Gc<'gc, 'own, Container<'gc, 'own>>>);

unsafe impl<'gc, 'own> Trace<'own> for Container<'gc, 'own> {
    type Gc<'to> = Container<'to, 'own>;

    fn needs_trace() -> bool
    where
        Self: Sized,
    {
        true
    }

    fn trace(&self, marker: Marker<'own, '_>) {
        self.0.trace(marker)
    }
}

fn main() {
    dreck!(owner1, arena1);
    dreck!(owner2, arena2);

    let mut container = Container(None);
    let ptr = arena1.add(container);

    let guard = pin!(RootGuard::new());
    let _ptr = root!(&arena2, guard, ptr);

    arena1.collect(&owner1);
    arena2.collect(&owner2);
}
