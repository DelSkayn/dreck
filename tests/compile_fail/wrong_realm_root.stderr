warning: variable does not need to be mutable
  --> tests/compile_fail/wrong_realm_root.rs:25:9
   |
25 |     let mut container = Container(None);
   |         ----^^^^^^^^^
   |         |
   |         help: remove this `mut`
   |
   = note: `#[warn(unused_mut)]` on by default

error[E0597]: `_pin` does not live long enough
  --> tests/compile_fail/wrong_realm_root.rs:23:5
   |
23 |     dreck!(owner2, arena2);
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     borrowed value does not live long enough
   |     binding `_pin` declared here
...
33 | }
   | -
   | |
   | `_pin` dropped here while still borrowed
   | borrow might be used here, when `_lifetime_constrainer` is dropped and runs the `Drop` code for type `main::KeepTillScopeDrop`
   |
   = note: values in a scope are dropped in the opposite order they are defined
   = note: this error originates in the macro `dreck` (in Nightly builds, run with -Z macro-backtrace for more info)
